### DISCUSSION TITLE
Ethernet MAC Address

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/STM32F7xx#17] [General]
The current ST code always sets the same MAC address, which causes issues when multiple boards are present.

I've just [committed](https://github.com/dresco/STM32H7xx/commit/a515501f9d3b7cdabf3164cc0412fcf0fa110272) a branch to the H7 driver, that I think takes a better approach (I borrowed the idea from the Zephyr codebase).

First three bytes are the ST OUI prefix, but with the LAA bit set.
Last three bytes from a CRC32 hash of the 96 bit chip UID.

I think this is a reasonable approach to generating persistently unique IDs, what do you think? If you are in favour, I could create a PR for the F7 repo also..

**Comment by terjeio** [comment]
> I think this is a reasonable approach to generating persistently unique IDs, what do you think?

It should be kind of ok? IIRC not strictly according to the rules though...

FYI for the Wiznet drivers I have added a [setting for MAC address](https://github.com/grblHAL/Plugin_networking/blob/0109e3e39d590b94057fd12d0cfddf8f4bb75f6c/wiznet/enet.c#L582), so one can get a "legal" one from a scrapped device and use that. Perhaps add it to the F7/H7 drivers and populate it with a default value from your algorithm making it overridable?

**Reply by dresco** [reply]
> It should be kind of ok? IIRC not strictly according to the rules though...

I think setting the LAA bit is the key point, as then only has to be locally unique (like auto-generated VM MAC addresses etc). But yes, agree it's a fudge for not having a legitimate hardware address.

> FYI for the Wiznet drivers I have added a [setting for MAC address](https://github.com/grblHAL/Plugin_networking/blob/0109e3e39d590b94057fd12d0cfddf8f4bb75f6c/wiznet/enet.c#L582), so one can get a "legal" one from a scrapped device and use that. Perhaps add it to the F7/H7 drivers and populate it with a default value from your algorithm making it overridable?

That sounds ideal, will do that.. Thanks!

**Reply by dresco** [reply]
Is there a reason the MAC address is printed in reverse here? 

https://github.com/grblHAL/Plugin_networking/blob/0109e3e39d590b94057fd12d0cfddf8f4bb75f6c/networking.c#L119-L129

**Comment by dresco** [comment]
Thanks @terjeio, updating the H7 code in line with yours.

One thing, I don't see `HAS_MAC_SETTING` being defined anywhere in the repo?

https://github.com/grblHAL/core/blob/d4c9d21ab2fe056bfaecef9471d77cb23a319590/plugins.h#L95

**Reply by terjeio** [reply]
> I don't see HAS_MAC_SETTING being defined anywhere in the repo?

Not yet, to be defined in a .pio file or build configuration if needed?

**Comment by dresco** [comment]
Thanks
